server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java

```diff
@@ -0,0 +1,168 @@
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.stream.IntStream;
+
+import static io.crate.metadata.functions.TypeVariableConstraint.typeVariable;
+import static io.crate.types.DataTypes.INTEGER;
+import static io.crate.types.TypeSignature.parseTypeSignature;
+
+public final class GenerateSubscripts<T> extends TableFunctionImplementation<T> {
+
+    public static final FunctionName NAME = new FunctionName(PgCatalogSchemaInfo.NAME, "generate_subscripts");
+    private static final RowType RETURN_TYPE = new RowType(List.of(INTEGER), List.of("subscripts"));
+    private static final Iterator<Row> NO_RESULTS_ITERATOR = Collections.emptyIterator();
+    private static final Comparator<Integer> KEEP_ORDER = (o1, o2) -> 0;
+
+
+    public static void register(TableFunctionModule module) {
+        module.register(
+            Signature.table(
+                NAME,
+                parseTypeSignature("array(E)"),
+                DataTypes.INTEGER.getTypeSignature(),
+                DataTypes.INTEGER.getTypeSignature()
+            ).withTypeVariableConstraints(typeVariable("E")),
+            GenerateSubscripts::new
+        );
+        module.register(
+            Signature.table(
+                NAME,
+                parseTypeSignature("array(E)"),
+                DataTypes.INTEGER.getTypeSignature(),
+                DataTypes.BOOLEAN.getTypeSignature(),
+                DataTypes.INTEGER.getTypeSignature()
+            ).withTypeVariableConstraints(typeVariable("E")),
+            GenerateSubscripts::new
+        );
+    }
+
+    private final Signature signature;
+    private final Signature boundSignature;
+
+    private GenerateSubscripts(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    private static List<?> extractNestedArray(List<?> array, int targetLevel) {
+        List<?> nested = array;
+        for (int level = 2; level <= targetLevel; level++) {
+            if (nested == null || nested.isEmpty()) {
+                return null;
+            }
+            if (!nested.stream().allMatch(e -> e == null || e instanceof List)) {
+                throw new IllegalArgumentException("expected a nested array at level " + level);
+            }
+            List<?> head = (List<?>) nested.get(0);
+            if (head == null || head.isEmpty()) {
+                return null;
+            }
+            if (!nested.stream().allMatch(e -> e == null || ((List<?>) e).size() == head.size())) {
+                throw new IllegalArgumentException("nested arrays should have the same dimension within a level");
+            }
+            nested = head;
+        }
+        return nested;
+    }
+
+    @Override
+    public Iterable<Row> evaluate(TransactionContext txnCtx, Input<T>... args) {
+        assert args.length == 2 || args.length == 3 : "number of args must be 2 or 3";
```

Comment:
not really related, but i am wondering whether assertion like that still makes sense. isn't it already guarded by the func resolving/registration logic?

Reply:
It is guarded by the function signature, but using assertions like this can be a useful defensive programming technique. 
It makes explicit that there exists some contract. It can help debugging in case the other component no longer ensures this.

But the assertion message could be improved. 

```suggestion
        assert args.length == 2 || args.length == 3 : "Signature must ensure that there are either two or three arguments"; 
```